src/main/scala/rise/core/equality.scala

```diff
@@ -0,0 +1,183 @@
+package rise.core
+
+import rise.core.semantics._
+import rise.core.types._
+import util.PatternMatching
+
+object equality {
+  case class Env[T](unwrap: Map[T, T]) {
+    def add(x: T, y: T): Env[T] = Env(unwrap.updated(x, y))
+    def check(x: T, y: T): Boolean = x == y || unwrap.get(x).contains(y)
+  }
+  object Env {
+    def apply[T](): Env[T] = Env(Map())
+  }
+
+  val equivNat: Env[Kind.Identifier] => Nat => Nat => Boolean = env => a => b => {
+    val natEnv = env.unwrap.collect { case (i: NatIdentifier, n: Nat) => (i, n) }
+    // substitutes elements on the left with elements on the right
+    substitute.natsInNat(natEnv, a) == b
+  }
+
+  trait TypeEq {
+    final type Eq[K <: Kind] = K#T => K#T => Boolean
+    def apply[K <: Kind] : Eq[K] = equiv[K](Env())
+    def equiv[K <: Kind] : Env[Kind.Identifier] => Eq[K]
+    def hash[K <: Kind] : K#T => Int
+  }
+
+  object typeErasure extends TypeEq {
+    override def hash[K <: Kind]: K#T => Int = _ => 0
+    override def equiv[K <: Kind]: Env[Kind.Identifier] => Eq[K] = _ => _ => _ => true
+  }
+
+  object typeUnificationAlphaEq extends TypeEq {
+    override def hash[K <: Kind]: K#T => Int = _ => 0
+    override def equiv[K <: Kind]: Env[Kind.Identifier] => Eq[K] = env => a => b => (a, b) match {
+      case (a : Type, b : Type) => (a, b) match {
+        case (TypePlaceholder, _) => true
+        case (_, TypePlaceholder) => true
+        case _ => typeAlphaEq.equiv[TypeKind](env)(a)(b)
+      }
+      case _ => typeAlphaEq.equiv(env)(a)(b)
+    }
+  }
+
+  object typeAlphaEq extends TypeEq {
+    /** Alpha equivalence on types.
+      * Short circuits in the event of syntactic equality.
+      * Kind equality is checked on dependent functions and pairs.
+      */
+    override def equiv[K <: Kind]: Env[Kind.Identifier] => Eq[K] = env => a => b => {
+      val and = PatternMatching.matchWithDefault(b, false)
+      a == b || (a match {
+        case a : Nat => and {case b : Nat => equivNat(env)(a)(b)}
+        case a : Type => and {case b : Type => equivType(env)(a)(b) }
+        case ia: Kind.Identifier => and { case ib: Kind.Identifier => env.check(ia, ib) }
+        case a: AddressSpace => and { case b: AddressSpace => (a : AddressSpace) == (b : AddressSpace) }
+        case NatToNatLambda(na, ba) => and { case NatToNatLambda(nb, bb) => equivNat(env.add(na, nb))(ba)(bb) }
+        case NatToDataLambda(na, ba) => and { case NatToDataLambda(nb, bb) => equiv[DataKind](env.add(na, nb))(ba)(bb) }
+        case NatCollectionFromArray(a) => and { case NatCollectionFromArray(b) => a == b } // FIXME: should use exprEq
+      })
+    }
+    val equivType: Env[Kind.Identifier] => Type => Type => Boolean = env => a => b => {
+      val and = PatternMatching.matchWithDefault(b, false)
+      a == b || (a match {
+        // Base cases
+        case TypePlaceholder => and { case TypePlaceholder => true }
```

Comment:
I think we should keep it as is since `TypePlaceholder`s are `Type`s. Ideally this wouldn't be the case, we would keep `Option[Type]` where relevant and be able to talk of `Type`s without `TypePlaceholder`.

Reply:
(Also, backwards compatibility: this was the old behavior as well.)